pkg/server/plugin/nodeattestor/k8s/sat.go

```diff
@@ -0,0 +1,307 @@
+	ServiceAccountWhitelist []string `hcl:"service_account_whitelist"`
+}
+
+type SATAttestorConfig struct {
+	Clusters map[string]*ClusterConfig `hcl:"clusters"`
+}
+
+type clusterConfig struct {
+	name               string
+	serviceAccountKeys []crypto.PublicKey
+	serviceAccounts    map[string]bool
+}
+
+type satAttestorConfig struct {
+	trustDomain string
+	clusters    []*clusterConfig
+}
+
+type SATAttestorPlugin struct {
+	mu     sync.RWMutex
+	config *satAttestorConfig
+}
+
+var _ nodeattestor.Plugin = (*SATAttestorPlugin)(nil)
+
+func NewSATAttestorPlugin() *SATAttestorPlugin {
+	return &SATAttestorPlugin{}
+}
+
+func (p *SATAttestorPlugin) Attest(stream nodeattestor.Attest_PluginStream) error {
+	req, err := stream.Recv()
+	if err != nil {
+		return satError.Wrap(err)
+	}
+
+	config, err := p.getConfig()
+	if err != nil {
+		return err
+	}
+
+	if req.AttestedBefore {
+		return satError.New("node has already attested")
+	}
+
+	if req.AttestationData == nil {
+		return satError.New("missing attestation data")
+	}
+
+	if dataType := req.AttestationData.Type; dataType != pluginName {
+		return satError.New("unexpected attestation data type %q", dataType)
+	}
+
+	if req.AttestationData.Data == nil {
+		return satError.New("missing attestation data payload")
+	}
+
+	attestationData := new(k8s.SATAttestationData)
+	if err := json.Unmarshal(req.AttestationData.Data, attestationData); err != nil {
+		return satError.New("failed to unmarshal data payload: %v", err)
+	}
+
+	if attestationData.UUID == "" {
+		return satError.New("missing UUID in attestation data")
+	}
+
+	if attestationData.Token == "" {
+		return satError.New("missing token in attestation data")
+	}
+
+	token, err := jwt.ParseSigned(attestationData.Token)
+	if err != nil {
+		return satError.New("unable to parse token: %v", err)
+	}
+
+	cluster, claims, err := verifyTokenSignature(config.clusters, token)
+	if err != nil {
+		return err
+	}
+
+	// TODO: service account tokens don't currently expire.... when they do, validate the time (with leeway)
+	if err := claims.Validate(jwt.Expected{
+		Issuer: "kubernetes/serviceaccount",
+	}); err != nil {
+		return satError.New("unable to validate token claims: %v", err)
+	}
+
+	if claims.Namespace == "" {
+		return satError.New("token missing namespace claim")
+	}
+
+	if claims.ServiceAccountName == "" {
+		return satError.New("token missing service account name claim")
+	}
+
+	serviceAccountName := fmt.Sprintf("%s:%s", claims.Namespace, claims.ServiceAccountName)
+
+	if !cluster.serviceAccounts[serviceAccountName] {
+		return satError.New("%q is not a whitelisted service account", serviceAccountName)
+	}
+
+	return stream.Send(&nodeattestor.AttestResponse{
+		Valid:        true,
+		BaseSPIFFEID: k8s.AgentID(config.trustDomain, attestationData.UUID),
+		Selectors: []*common.Selector{
+			makeSelector("cluster:name", cluster.name),
+			makeSelector("service-account:namespace", claims.Namespace),
```

Comment:
the plugin name (and hence beginning of all selectors) is `k8s_sat` which makes me lean towards snake case for service account too

`k8s_sat:service_account`